homeassistant/components/recorder/purge.py

```diff
@@ -101,3 +59,67 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
     except SQLAlchemyError as err:
         _LOGGER.warning("Error purging history: %s", err)
     return True
+
+
+def _select_event_ids_to_purge(session, purge_before):
+    """Return a list of event ids to purge."""
+    events = (
+        session.query(Events.event_id)
+        .filter(Events.time_fired < purge_before)
+        .limit(MAX_ROWS_TO_PURGE)
+        .all()
+    )
+    _LOGGER.debug("Selected %s event ids to remove", len(events))
+    return [event.event_id for event in events]
+
+
+def _select_state_ids_to_purge(session, event_ids):
+    """Return a list of state ids to purge."""
+    states = session.query(States.state_id).filter(States.event_id.in_(event_ids)).all()
```

Comment:
Personally, I don't really like the constant `in_` checks (here and in the functions below). Unfortunately, they are the only way to do it, but very inefficient. Does it work/would it make sense to do some pre-filtering? Maybe with the assumption
```py
.filter(States.last_updated < purge_before)
```

Reply:
Since its the primary key the query optimizer is going to drop the `purge_before

```
MariaDB [homeassistant]> explain select * from events where event_id IN (504944, 504945) and time_fired > '2021-03-06 10:16:28';
+------+-------------+--------+-------+------------------------------+---------+---------+------+------+-------------+
| id   | select_type | table  | type  | possible_keys                | key     | key_len | ref  | rows | Extra       |
+------+-------------+--------+-------+------------------------------+---------+---------+------+------+-------------+
|    1 | SIMPLE      | events | range | PRIMARY,ix_events_time_fired | PRIMARY | 4       | NULL |    2 | Using where |
+------+-------------+--------+-------+------------------------------+---------+---------+------+------+-------------+
1 row in set (0.001 sec)
```

Reply:
Just saw that. I wanted to use it on the `states` queries.
```
> explain select * from states where states.event_id IN (504944, 504945) and states.last_updated > '2021-03-06 10:16:28'

+------+-------------+--------+--------------+-------------------------------------------+-------------------------------------------+-----------+------+------+--------------------------------------------------------+
| id   | select_type | table  | type         | possible_keys                             | key                                       | key_len   | ref  | rows | Extra                                                  |
+------+-------------+--------+--------------+-------------------------------------------+-------------------------------------------+-----------+------+------+--------------------------------------------------------+
|    1 | SIMPLE      | states | range|filter | ix_states_last_updated,ix_states_event_id | ix_states_last_updated,ix_states_event_id | 6|5       | NULL |    1 | Using index condition; Using where; Using rowid filter |
+------+-------------+--------+--------------+-------------------------------------------+-------------------------------------------+-----------+------+------+--------------------------------------------------------+
```

Reply:
done